crates/sui-graphql-rpc/src/data/move_registry_data_loader.rs

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+
+use async_graphql::dataloader::{DataLoader, Loader};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    config::MoveRegistryConfig,
+    error::Error,
+    types::{
+        base64::Base64,
+        dot_move::{
+            error::MoveRegistryError,
+            on_chain::{AppRecord, Name},
+        },
+    },
+};
+
+/// GraphQL fragment to query the values of the dynamic fields.
+const QUERY_FRAGMENT: &str =
+    "fragment RECORD_VALUES on DynamicField { value { ... on MoveValue { bcs } } }";
+
+pub(crate) struct ExternalNamesLoader {
+    client: reqwest::Client,
+    config: MoveRegistryConfig,
+}
+
+/// Helper types for accessing a shared `DataLoader` instance.
+#[derive(Clone)]
+pub(crate) struct MoveRegistryDataLoader(pub Arc<DataLoader<ExternalNamesLoader>>);
+
+impl ExternalNamesLoader {
+    pub(crate) fn new(config: MoveRegistryConfig) -> Self {
+        Self {
+            client: reqwest::Client::new(),
+            config,
+        }
+    }
+
+    /// Constructs the GraphQL Query to query the names on an external graphql endpoint.
+    fn construct_names_graphql_query(&self, names: &[Name]) -> (String, HashMap<Name, usize>) {
+        let mut mapping: HashMap<Name, usize> = HashMap::new();
+
+        let mut result = format!(r#"{{ owner(address: "{}") {{"#, self.config.registry_id);
+
+        // we create the GraphQL query keys with a `fetch_{id}` prefix, which is accepted on graphql fields.
+        for (index, name) in names.iter().enumerate() {
+            let bcs_base64 = name.to_base64_string();
+
+            // retain the mapping here (id to bcs representation, so we can pick the right response later on)
+            mapping.insert(name.clone(), index);
+
+            result.push_str(&format!(
+                r#"{}: dynamicField(name: {{ type: "{}::name::Name", bcs: {} }}) {{ ...RECORD_VALUES }}"#,
+                fetch_key(&index),
+                self.config.package_address,
+                bcs_base64
+            ));
```

Comment:
Here is a playground link demonstrating what I mentioned about using `write!` directly before:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=66009e35c7c6d3e53173a813d5531ebd

The key is that you need to have `std::fmt::Write` in scope, and you are writing to the mutable borrow of the string.

Reply:
Ahh you're right... That's what it was complaining for (missing `std::fmt::Write`). Sorry for having you say it to me one million times 🫠